rust/kernel/net/tcp/cong.rs

```diff
@@ -0,0 +1,544 @@
+    fn set_state(sk: &mut Sock<Self>, new_state: State) {}
+
+    /// Called when removing ACKed packets from the retransmission queue. Can be
+    /// used for packet ACK accounting.
+    fn pkts_acked(sk: &mut Sock<Self>, sample: &AckSample) {}
+
+    /// Called to undo a recent cwnd reduction that was found to has been
+    /// unnecessary. Returns the new value of cwnd.
+    fn undo_cwnd(sk: &mut Sock<Self>) -> u32;
+
+    /// Initializes the private data.
+    ///
+    /// When this function is called, [`sk.inet_csk_ca()`] will contain a value
+    /// returned by `Self::Data::default()`.
+    ///
+    /// Only implement this function when you need to perform additional setup
+    /// tasks.
+    ///
+    /// [`sk.inet_csk_ca()`]: Sock::inet_csk_ca
+    fn init(sk: &mut Sock<Self>) {}
+
+    /// Cleans up the private data.
+    ///
+    /// After this function returns, [`sk.inet_csk_ca()`] will be dropped.
+    ///
+    /// Only implement this function when you need to perform additional cleanup
+    /// tasks.
+    ///
+    /// [`sk.inet_csk_ca()`]: Sock::inet_csk_ca
+    fn release(sk: &mut Sock<Self>) {}
+}
+
+pub mod reno {
+    //! TCP Reno congestion control.
+    //!
+    //! Algorithms may choose to invoke these callbacks instead of providing
+    //! their own implementation.
+    use super::{Algorithm, Sock};
+    use crate::bindings;
+
+    #[inline]
+    pub fn undo_cwnd<T: Algorithm + ?Sized>(sk: &mut Sock<T>) -> u32 {
+        // SAFETY:
+        // - `sk` has been passed to the callback that invoked us,
+        // - it is OK to pass it to the callback of the Reno algorithm as it
+        //   will never touch the private data.
+        unsafe { bindings::tcp_reno_undo_cwnd(sk.sk) }
+    }
+}
+
+/// Representation of the `struct sock *` that is passed to the callbacks of the
+/// CCA.
+///
+/// Every callback receives a pointer to the socket that it is operating on.
+/// There are certain operations that callbacks are allowed to perform on the
+/// socket, and this type just exposes methods for performing those. This
+/// prevents callbacks from performing arbitrary manipulations on the socket.
+//  TODO: Currently all callbacks can perform all operations. However, this
+//  might be too permissive, e.g., the `pkts_acked` callback should probably not
+//  be changing cwnd...
+///
+/// # Invariants
+///
+/// The wrapped `sk` must have been obtained as the argument to a callback of
+/// the congestion algorithm `T` (other than the `init` cb) and may only be used
+/// for the duration of that callback. In particular:
+///
+/// - `sk` points to a valid `struct sock`.
+/// - `tcp_sk(sk)` points to a valid `struct tcp_sock`.
+/// - The socket uses the CCA `T`.
+/// - `inet_csk_ca(sk)` points to a valid instance of `T::Data`, which belongs
+///   to the instance of the algorithm used by this socket. A callback has
+///   exclusive, mutable access to this data.
+pub struct Sock<T: Algorithm + ?Sized> {
```

Comment:
This struct lives in `net/sock.h` right? It should be in the same module here.

I feel like the C side sock vs. socket might not be something we want to keep, but I don't have any better suggestion.

Making this type generic over a CCA algorithm seems somewhat specialized.